mgr/dashboard: fix rgw service creation in dashboard

aaSharma14 opened the pull request:
In the rgw service creation form provide an option to add the realm, zonegroup and zone name in separate fields rather than entering a pattern in the service id

https://tracker.ceph.com/issues/61252

[Screencast from 2023-05-31 18-56-07.webm](https://github.com/ceph/ceph/assets/66050535/98319a89-f0bf-44af-b2ee-505ffd91ae03)








<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

Pegonzal reviewed (commented):
@aaSharma14 unit test failing over here

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -83,15 +83,54 @@
+          </div>
+        </div>
+
+        <div class="form-group row"
+             *ngIf="serviceForm.controls.service_type.value === 'rgw'">
+          <label class="cd-col-form-label"
+                 for="zonegroup_name">
+            <span i18n>Zonegroup Name</span>
+          </label>
+          <div class="cd-col-form-input">
+            <input id="zonegroup_name"
+                   class="form-control"
+                   type="text"
+                   formControlName="zonegroup_name">
+          </div>
+        </div>
+
+        <div class="form-group row"
+             *ngIf="serviceForm.controls.service_type.value === 'rgw'">
+          <label class="cd-col-form-label"
+                 for="zone_name">
+            <span i18n>Zone Name</span>
+          </label>
+          <div class="cd-col-form-input">
+            <input id="zone_name"
+                   class="form-control"
+                   type="text"
+                   formControlName="zone_name">
+          </div>
```
Lets set the default value to "default" for zone and zonegroup, that's what is used.

cloudbehl replied:
Also add some helper text to each one of them

aaSharma14 replied:
Done

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:
```diff
@@ -559,12 +553,16 @@ export class ServiceFormComponent extends CdForm implements OnInit {
-      if (svcIdMatch[3]) {
-        serviceSpec['rgw_realm'] = svcIdMatch[3];
-        serviceSpec['rgw_zone'] = svcIdMatch[4];
+      if (
+        values['realm_name'] !== null &&
+        values['zone_name'] !== null &&
+        values['zonegroup_name'] !== null
+      ) {
```
why do we have a `&&` condition, even one value is set we want to reflect that in service.

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:
```diff
@@ -559,12 +553,16 @@ export class ServiceFormComponent extends CdForm implements OnInit {
-      if (svcIdMatch[3]) {
-        serviceSpec['rgw_realm'] = svcIdMatch[3];
-        serviceSpec['rgw_zone'] = svcIdMatch[4];
+      if (
+        values['realm_name'] !== null &&
+        values['zone_name'] !== null &&
+        values['zonegroup_name'] !== null
+      ) {
+        serviceSpec['rgw_realm'] = values['realm_name'];
+        serviceSpec['rgw_zonegroup'] = values['zonegroup_name'];
+        serviceSpec['rgw_zone'] = values['zone_name'];
       }
```
```suggestion
        serviceSpec['rgw_realm'] = values?.realm_name || null;
        serviceSpec['rgw_zonegroup'] = values?.zonegroup_name || null];
        serviceSpec['rgw_zone'] = values?.zone_name || null;
```

Pegonzal reviewed (commented):
I'm not a fan of the current implementation of the create realm, zonegroup and zone from the form

The text and the button are wierdly placed.
The text is redundant, `You can create a realm/zonegroup/zone by clicking here` next to the `Create Realm/Zonegroup/Zone` button.
The final text says: `By default the default realm/zonegroup/zone will be selected` which I also feel is redundant. If we are already filling the form fields with the default realm, zonegroup and zone is it really necessary the text?

A part from that nowhere is specified that the realm, zgp and zn we are creating will be set as the default ones.

Maybe something like:
![image](https://github.com/ceph/ceph/assets/106810838/1192ebfe-ae29-438c-9b1b-99a0a80fbdbd)

Or directly giving the option to type in the field and create it if it doesnt exist would work better, like we have in other places such as creating an alert silence

For some reason my zonegroup and zones fields are not being filled by default
[screen-capture (10).webm](https://github.com/ceph/ceph/assets/106810838/3243c119-8a46-49c6-9901-9cea5cc874ed)

Also after creating the realm, zonegroup and zone
[screen-capture (11).webm](https://github.com/ceph/ceph/assets/106810838/97071f36-8562-4e0e-8378-c1844f511a6a)

aaSharma14 commented on src/pybind/mgr/dashboard/controllers/rgw.py:
```diff
@@ -126,6 +127,184 @@ def migrate(self, realm_name=None, zonegroup_name=None, zone_name=None,
+            all_zones_info['default_zone'] = rgw_zone_list['default_info']  # type: ignore
+        else:
+            all_zones_info['default_zone'] = ''  # type: ignore
+        return all_zones_info
```
Updated the PR

aaSharma14 commented:
> I'm not a fan of the current implementation of the create realm, zonegroup and zone from the form
> 
> The text and the button are wierdly placed. The text is redundant, `You can create a realm/zonegroup/zone by clicking here` next to the `Create Realm/Zonegroup/Zone` button. The final text says: `By default the default realm/zonegroup/zone will be selected` which I also feel is redundant. If we are already filling the form fields with the default realm, zonegroup and zone is it really necessary the text?
> 
> A part from that nowhere is specified that the realm, zgp and zn we are creating will be set as the default ones.
> 
> Maybe something like: ![image](https://user-images.githubusercontent.com/106810838/241925214-1192ebfe-ae29-438c-9b1b-99a0a80fbdbd.png)
> 
> Or directly giving the option to type in the field and create it if it doesnt exist would work better, like we have in other places such as creating an alert silence
> 
> For some reason my zonegroup and zones fields are not being filled by default [screen-capture (10).webm](https://github.com/ceph/ceph/assets/106810838/3243c119-8a46-49c6-9901-9cea5cc874ed)
> 
> Also after creating the realm, zonegroup and zone [screen-capture (11).webm](https://github.com/ceph/ceph/assets/106810838/97071f36-8562-4e0e-8378-c1844f511a6a)

Thanks @Pegonzal . I have addressed most of the comments
